packages/autocomplete-core/src/utils/createCancelablePromise.ts

```diff
@@ -0,0 +1,160 @@
+import { noop } from '@algolia/autocomplete-shared';
+
+type CancelablePromiseState = {
+  isCanceled: boolean;
+  onCancelList: Array<(...args: any[]) => any>;
+};
+
+type PromiseExecutor<TValue> = (
+  resolve: (value: TValue | PromiseLike<TValue>) => void,
+  reject: (reason?: any) => void
+) => void;
+
+type CreateInternalCancelablePromiseParams<TValue> = {
+  executor?: PromiseExecutor<TValue>;
+  promise?: Promise<TValue>;
+  initialState?: CancelablePromiseState;
+};
+
+function createInternalCancelablePromise<TValue>({
```

Comment:
with this implementation it's technically a thenable, not a promise. I don't think it really matters though.

Making it a real promise and adding the functions dynamically could be interesting, but likely makes the code more complex

Reply:
Yeah, it's a `PromiseLike`. I think it's fine, we have exactly what we need and we understand what it does.

Later on, if we want to reuse it, it could be interesting to externalize it as a modular implementation—you piece your own and specify the methods you want.